### PR TITLE
Falling nodes: Add support for facedir, colorfacedirr, wallmounted, colorwallmounted, color, airlike, signlike, torchlike and glow

### LINKED ISSUE #5324: Falling nodes of “node box” type don't rotate properly
If you have a falling node from a node with a nodebox appearance with a nodebox which can rotate (thanks to `facedir`), the falling node is sometimes incorrectly rotated when it falls.
Example: If you have an anvil node box, and it drops, for some orientations, the anvil gets rotated back to its default orientation while it falls. When it reaches the bottom and becomes a node again, the orientation is OK again.

Tested in Minetest 0.4.15.
I discovered this bug while developing MineClone 2. To test this, you can try to place an anvil from the creative inventory in MineClone 2 version 0.13.0 and make it fall. Test it with different orientations.

My guess: Minetest simply ignores the facedir when rendering the falling node.

### PR DESCRIPTION
**Pull request by Wuzzy2**
## Goal of the PR

Falling nodes are currently not supporting rotation (`facedir`, `wallmounted`) nor node colorization. This PR fixes this. There are also other rendering-related features.

Fixes #5324.

## Features

Support for these paramtype2's is added:

* `facedir`
* `colorfacedir`
* `wallmounted`
* `colorwallmounted`
* `color`

This means, falling nodes will now be properly rotated and colorized.

Falling nodes will get improved support for the following drawtypes:

* `airlike`: Nothing will render (will also be non-pointable due to how the `is_visible` property works)
* `signlike`: Fix rotation, use simple sprite
* `torchlike`: Use standing torch for dropping

Finally, `glow` is applied to the entity if the node is a `light_source`.

## To do

I think it's ready, but I urge core devs to do their own testing. Well, I did a lot of testing already, but you probably want to do you own testing since this PR touches a very sensitive part of the game.

## How to test

This is the hard part. You are going to need a LOT of nodes. To test this, you would need nodes with the properties I just listed above and some easy way to change param2 as well. And then you just make each node with “every” possible param2 value fall.

I have actually written test code to make testing this easier, but it's not released yet. I will post in this thread again when my test code is public.

### CODE REVIEW

**Comment by Wuzzy2** [comment]
Okay, as promised, here are instructions on how to test this:

1. Download DevToys modpack here: https://forum.minetest.net/viewtopic.php?f=9&t=23802
2. Use it in MTG
3. In the Creative Inventory, pick up the various test nodes (those that are relevant to this PR, i.e. Color Test Node, Facedir Test Node, etc.). Place them in the world
4. Punch the nodes with the Falling Node Tool to make them fall. Check if they behave as promised
5. Use the Param2 Tool to change the `param2` values for the nodes. You want to test the falling nodes under various different `param2` values
6. Feel free to also test the behavior with custom nodes and make them fall with the Falling Node Tool

Targets:
* The facedir, wallmounted and colored nodes MUST be correctly rotated AND correctly colored for valid values of `param2`
* Light source nodes must glow when falling (test in darkness)
* The Airlike Test Node must be invisible when falling
* Signlike Test Node must be correctly rotated (note this is not perfect yet, there's a weird offset)
* Torchlike Test Node must be in “standing up” position during falling (this design choice is deliberate)

Note I have already done extensive testing myself. But it's better to double-check. :)

File: builtin/game/falling.lua [hunk 1]
```diff
@@ -33,11 +33,103 @@ core.register_entity(":__builtin:falling_node", {
...
-			textures = {node.name},
-		})
+		if def.drawtype == "airlike" then
+			self.object:set_properties({
+				is_visible = false,
+			})
+		else
+			local glow
+			if def then
+				glow = def.light_source
```
Comment (MoNTE48) [inline builtin/game/falling.lua]: The same :)

**Comment by SmallJoker** [comment]
How are these calculations different to 
 - `minetest.facedir_to_dir` - for `facedir`, `colorfacedir`
 - `minetest.wallmounted_to_dir` - for `wallmounted`, `colorwallmounted`
?

File: builtin/game/falling.lua [hunk 2]
```diff
@@ -33,11 +33,105 @@ core.register_entity(":__builtin:falling_node", {
...
-		self.object:set_properties({
-			is_visible = true,
-			textures = {node.name},
-		})
+		if def.drawtype == "airlike" then
+			self.object:set_properties({
+				is_visible = false,
+			})
+		else
+			if def.drawtype == "torchlike" or def.drawtype == "signlike" then
```
Comment (SmallJoker) [inline builtin/game/falling.lua]: `elseif def.drawtype == "torchlike" or def.drawtype == "signlike" then`
Daisy-chain style has the same efficiency and reduces the amount of indents at the same time.

File: builtin/game/falling.lua [hunk 3]
```diff
@@ -33,11 +33,105 @@ core.register_entity(":__builtin:falling_node", {
...
+					is_visible = true,
+					wield_item = itemstring,
+					glow = def.light_source,
+				})
+			end
+		end
+		-- Rotate entity
+		if def.drawtype == "torchlike" then
+			self.object:set_yaw(math.pi*0.25)
+		elseif (node.param2 ~= 0 and (def.wield_image == "" or def.wield_image == nil)) or def.drawtype == "signlike" then
```
Comment (SmallJoker) [inline builtin/game/falling.lua]: 115 characters (after tabulators). Please break it into two or three lines to keep it <= 80 characters.

File: builtin/game/falling.lua [hunk 4]
```diff
@@ -33,11 +33,105 @@ core.register_entity(":__builtin:falling_node", {
 				end
 			end
 		end
+		local def = core.registered_nodes[node.name]
+		if not def then
+			-- Don't allow unknown nodes to fall
+			minetest.log("warning", "Unknown falling node removed at "..minetest.pos_to_string(self.object:get_pos()))
```
Comment (SmallJoker) [inline builtin/game/falling.lua]: Line too long here as well.

**Comment by Wuzzy2** [comment]
> How are these calculations different to (...)

I don't know which calculations you are referring to specifically.

But well, the entity rotation needs pitch, yaw and roll. And the functions you have provided don't do that. Remember entity rotation is a little bit “special”, so these calculations are neccessary.

I don't think we have converter functions to convert from facedir/wallmounted to entity rotation.

**Comment by Wuzzy2** [comment]
FYI: This is ready for testing. I do not plan to add new features.

**Comment by SmallJoker** [comment]
Sorry for the late answer.

> I don't know which calculations you are referring to specifically.

Mainly those two:
https://github.com/minetest/minetest/blob/7c0e0c34ecb66e044f59cbab03a4c12fc7838e97/doc/lua_api.txt#L4763-L4765
https://github.com/minetest/minetest/blob/7c0e0c34ecb66e044f59cbab03a4c12fc7838e97/doc/lua_api.txt#L4769-L4771

I know that rotation is different to direction, but this can be converted to avoid code duplication. Hopefully performance is still negligible with this approach.

```Lua
local dir = minetest.facedir_to_dir(node.param2) -- or wallmounted_to_dir

local yaw = math.atan2(-dir.x, dir.z) -- From dev.minetest.net
-- Order and signs might not be correct. Not tested
local pitch = math.atan2(dir.x, -dir.y)
local roll = math.atan2(-dir.z, dir.y)
self.object:set_rotation({x=pitch, y=yaw, z=roll})
```

**Comment by Wuzzy2** [comment]
Hmmm, I don't know if it's a good idea to use `atan2`. What if a ton of falling nodes want to spawn at once?

Maybe a lookup-table-based approach is better. Using `*_to_dir` to get the `dir`, then use `dir` to lookup angles.

While I figure out how to proceed, are there other problems in this PR?

**Comment by SmallJoker** [comment]
@Wuzzy2 Moreover I'm only concerned about `core.log("warning", ...)` where a bunch of unknown nodes could spam the logs (which is definitely slower than `atan2`).

**Comment by Wuzzy2** [comment]
The unknown node warning is actually unrelated to this PR … Do you want to reduce the log level to `info`?

**Comment by Wuzzy2** [comment]
Oops, I just realized it was my fault indeed. OK, log level was reduced to `info`.

Well, about lookup vs atan2: In both cases you have to touch x, y and z values, so I don't see a difference in principle.

**Comment by Wuzzy2** [comment]
@SmallJoker: Your suggested approach is not working at all. `minetest.facedir_to_dir` does not give us enough information to unambigiously rotate the node properly. I tried.
Frankly, I wonder if this additonal conversion is even needed. My code is already functional and tested, and it's fast. Why tweak it?

So what *exactly* do you want me to do?

I am about to give up. This PR is finished, yet won't be accepted, and I have no idea how to proceed.

**Comment by Wuzzy2** [comment]
Okay, I have just moved the facedir handling code into a lookup table.

**Comment by SmallJoker** [comment]
> minetest.facedir_to_dir does not give us enough information to unambigiously rotate the node properly.

The entire point was to re-use builtin functions that already provide such conversions. You're right, I totally missed that it doesn't account all rotations (only 6). Sorry for the trouble.

Moving the conversion to a LUT gives it a better overview and will be easier to re-use later for an API function (if needed).

File: builtin/game/falling.lua [hunk 5]
```diff
@@ -33,11 +61,92 @@ core.register_entity(":__builtin:falling_node", {
...
+				visual_size = vsize,
+				glow = def.light_source,
+			})
+		end
+		-- Rotate entity
+		if def.drawtype == "torchlike" then
+			self.object:set_yaw(math.pi*0.25)
+		elseif (node.param2 ~= 0 and (def.wield_image == ""
+				or def.wield_image == nil))
+				or def.drawtype == "signlike" then
```
Comment (SmallJoker) [inline builtin/game/falling.lua]: What about nodebox drawtype? If you change `default:sign_wall_wood` to a falling node, the rotation is not correct.
EDIT: Neither are meshes -> example: `default:torch_wall`, the Y rotation is always off by 180°.

File: builtin/game/falling.lua [hunk 6]
```diff
@@ -33,11 +61,92 @@ core.register_entity(":__builtin:falling_node", {
...
+				glow = def.light_source,
+			})
+		else
+			local itemstring = node.name
+			if core.is_colored_paramtype(def.paramtype2) then
+				itemstring = core.itemstring_with_palette(itemstring, node.param2)
+			end
+			local vsize
+			if def.visual_scale then
+				vsize = { x = def.visual_scale * SCALE, y = def.visual_scale * SCALE, z = def.visual_scale * SCALE }
```
Comment (SmallJoker) [inline builtin/game/falling.lua]: Line is too long -- needs line breaks or substitution of repeated `def.visual_scale * SCALE`.

**Comment by Wuzzy2** [comment]
Nice catch, SmallJoker!

Wallmounted meshes, wallmounted nodeboxes, facedir meshes and facedir nodeboxes should all rotate properly now.
The overlong lines are shorter now.

A final question: Would it be better to switch to the `item` visual instead of `wielditem`?

File: builtin/game/falling.lua [hunk 7]
```diff
@@ -9,7 +37,7 @@ local builtin_shared = ...
 core.register_entity(":__builtin:falling_node", {
 	initial_properties = {
 		visual = "wielditem",
-		visual_size = {x = 0.667, y = 0.667},
+		visual_size = { x = SCALE, y = SCALE, z = SCALE },
```
Comment (MoNTE48) [inline builtin/game/falling.lua]: visual_size = {x = SCALE, y = SCALE, z = SCALE},
according to MT Code style guide

File: builtin/game/falling.lua [hunk 8]
```diff
@@ -1,6 +1,34 @@
 -- Minetest: builtin/item.lua
 
 local builtin_shared = ...
+local SCALE = 0.667
+
+local facedir_to_euler = {
```
Comment (MoNTE48) [inline builtin/game/falling.lua]: local pi = math.pi
will be a little faster

**Comment by Wuzzy2** [comment]
> local pi = math.pi
> will be a little faster

Is this actually true?

**Comment by MoNTE48** [comment]
Non-local: 0.719 (158%)
Localized: 0.453 (100%)

**Comment by Wuzzy2** [comment]
What's your methology?

File: builtin/game/falling.lua (hunk 8 above)
Comment (SmallJoker) [reply builtin/game/falling.lua]: Does not matter at all in this case because it's only run once during initialization.

**Comment by SmallJoker** [comment]
> A final question: Would it be better to switch to the item visual instead of wielditem?

`item` would fit better because the `wield_image` parameter is not wanted for nodes that are/were not wielded.
Will re-test soon.

**Comment by NathanSalapat** [comment]
Testing with the devtest game almost everything looks to be working correctly. The wallmounted nodes flip 180 degrees while falling, but then flip back to the original orientation when landing.

**Comment by Wuzzy2** [comment]
I believe you are using outdated versions. All wallmounted nodes work for me.

It would help if you could say the exact node name.

**Comment by NathanSalapat** [comment]
I downloaded the zip just before compiling and testing. I was using testnodes:attached_wallmounted. I just downloaded the devtest game before testing out the other branches.

**Comment by NathanSalapat** [comment]
Video demonstrations here: https://youtu.be/APIIx4Haw3A

**Comment by Wuzzy2** [comment]
Oh, you are right! This node was indeed dysfunctional. This should be fixed now.

I also fixed the rotation of signlike nodes, but not the center placement. The placement into the center cannot be improved atm. It seems there is no way to displace the sprite. So the center placement is an approximation and a neccessary evil. I will leave it as-is.

File: builtin/game/falling.lua [hunk 9]
```diff
@@ -33,11 +61,105 @@ core.register_entity(":__builtin:falling_node", {
...
-			is_visible = true,
-			textures = {node.name},
-		})
+		if def.drawtype == "airlike" then
+			self.object:set_properties({
+				is_visible = false,
+			})
+		elseif def.drawtype == "torchlike" or def.drawtype == "signlike" then
+			local textures
+			if def.tiles and def.tiles[1] then
```
Comment (sfan5) [inline builtin/game/falling.lua]: This is insufficient, def.tiles[1] could also be a table like:
`{name="image.png", backface_culling=bool, tileable_vertical=bool,
  tileable_horizontal=bool, align_style="node"/"world"/"user", scale=int}`

File: builtin/game/falling.lua [hunk 10]
```diff
@@ -33,11 +61,105 @@ core.register_entity(":__builtin:falling_node", {
...
+		end
 		self.meta = meta
-		self.object:set_properties({
-			is_visible = true,
-			textures = {node.name},
-		})
+		if def.drawtype == "airlike" then
+			self.object:set_properties({
+				is_visible = false,
+			})
```
Comment (sfan5) [inline builtin/game/falling.lua]: unnecessary IMO

File: builtin/game/falling.lua [hunk 11]
```diff
@@ -33,11 +61,105 @@ core.register_entity(":__builtin:falling_node", {
...
+				or def.wield_image == nil))
+				or def.drawtype == "signlike"
+				or def.drawtype == "mesh"
+				or def.drawtype == "normal"
+				or def.drawtype == "nodebox" then
+			if (def.paramtype2 == "facedir" or def.paramtype2 == "colorfacedir") then
+				local fdir = node.param2 % 32
+				-- Get rotation from a precalculated lookup table
+				local euler = facedir_to_euler[fdir + 1]
+				if euler then
```
Comment (sfan5) [inline builtin/game/falling.lua]: The table has 24 entries, `fdir` is [0, 32[
What are the missing 8 and why are the not included?

**Comment by Wuzzy2** [comment]
> The table has 24 entries, fdir is [0, 32[
> What are the missing 8 and why are the not included?

`facedir` only has 24 valid states. See `facedir` in `lua_api.txt`.
Color in `colorfacedir` changes every 32 values. So I conclude “missing” 8 other states are simply unused/undefined. Therefore, no special treatment in the code.

**Comment by Wuzzy2** [comment]
About `airlike` handling being “unneccessary”: No, it's not. If I don't do this, an `unknown_node.png` (IIRC) texture will appear.

> This is insufficient, def.tiles[1] could also be a table like:

OK, this is a valid complaint, it might even be a potential crasher. I need to think of a solution for this. Nice catch!

**Comment by Wuzzy2** [comment]
* `airlike`: OK, the code was indeed not needed, I looked closer. It has been removed now.
* Broken tiledef for `signlike`+`torchlike`: Fixed. It was indeed a crasher!
* `facedir`: No change, see above

**Comment by sfan5** [comment]
> airlike: OK, the code was indeed not needed

What I meant is that <i>support</i> for airlike is unneeded, nobody will ever want to use falling air.
But apparently the code was actually not needed, if it comes "for free" I won't insist on not supporting airlike.

File: builtin/game/falling.lua [hunk 12]
```diff
@@ -33,11 +61,109 @@ core.register_entity(":__builtin:falling_node", {
...
+			-- Do nothing
+		elseif def.drawtype == "torchlike" or def.drawtype == "signlike" then
+			local textures
+			if def.tiles and def.tiles[1] then
+				local tile
+				if type(def.tiles[1]) == "table" then
+					tile = def.tiles[1].name
+				else
+					tile = def.tiles[1]
+				end
```
Comment (sfan5) [inline builtin/game/falling.lua]: Sorry, the inefficiency here bothers me:
```lua
local tile = def.tiles[1]
if type(tile) == "table" then
    tile = tile.name
end
```

**Comment by Wuzzy2** [comment]
Done.

**Comment by paramat** [comment]
Build failure:
```
Checking builtin/game/falling.lua                 1 warning

    builtin/game/falling.lua:74:32: empty if branch
```

**Comment by Wuzzy2** [comment]
Ugh, what an annoying warning! But OK, it is fixed now.

**Pull request merged**